README.md
# ComS487 Network Programming
The projects in this course introduced me to the world of networking.
The projects in this repository are my implementations of network programs that utilize different protocols, sockets, and introduced me to multithreading.


hw1/CHANGELOG
commit 8fd095b1c84264ae9be821d8f0a0325c071836a2
Author: Akhil Sangem <yeesaurusrex@Yeesauruss-MacBook-Pro.local>
Date:   Mon Feb 5 03:32:53 2018 -0600

    Finalizing code. Formatting Code. Submitting.

commit b1c89a84f516ed72823f005d62ba55d7a84c20d3
Author: Akhil Sangem <yeesaurusrex@Yeesauruss-MacBook-Pro.local>
Date:   Mon Feb 5 02:39:31 2018 -0600

    Can now communicate both on udp and tcp
    Can now run both threads

commit bee1ad5e8356623fb6d6a822e7db0d7350ac3e7f
Author: Akhil Sangem <yeesaurusrex@Yeesauruss-MacBook-Pro.local>
Date:   Sun Feb 4 16:22:28 2018 -0600

    Able to establish tcp connection on manager and agent

commit 15fc6e0ddf8c327c605a4d413cbb14f3f60dd729
Author: Akhil Sangem <yeesaurusrex@Yeesauruss-MacBook-Pro.local>
Date:   Thu Feb 1 12:19:15 2018 -0600

    Edited struct order
    Can now send udp datagram and receive on manager.
    Manager can now store information from past agents in an array

commit 319e6ebfcfe43c538088f8ca106e1d68d749ebbb
Author: Akhil Sangem <yeesaurusrex@xyphos-macbookpro.student.iastate.edu>
Date:   Wed Jan 31 13:25:55 2018 -0600

    Set up a java server to recieve data from udp.

    Was able to successfully send a message.

commit 26a75ab6687c7e338a64ffb830214e039af52f99
Merge: e5fafdc a20b2d5
Author: Akhil Sangem <yeesaurusrex@Yeesauruss-MacBook-Pro.local>
Date:   Tue Jan 30 18:26:47 2018 -0600

    Merge branch 'master' of https://github.com/amorphos/ComS487

commit e5fafdc98706275c5ffa241937dadd97b578e55c
Author: Akhil Sangem <yeesaurusrex@Yeesauruss-MacBook-Pro.local>
Date:   Tue Jan 30 18:26:30 2018 -0600

    merging from mac

commit a20b2d5c9a54fb3a0d598ef57f23ab48b129907a
Merge: 4278893 b430313
Author: Akhil Sangem <akhilsangem@gmail.com>
Date:   Tue Jan 30 17:12:20 2018 -0600

    Merge branch 'master' of https://github.com/amorphos/ComS487

commit 4278893465ff2553786592a2d7f7267843f49cda
Author: Akhil Sangem <akhilsangem@gmail.com>
Date:   Tue Jan 30 17:12:01 2018 -0600

    Added code for sending a UDP datagram

commit b43031314a33c93daae59f373c03005e61f07560
Author: Akhil Sangem <akhilsangem@gmail.com>
Date:   Tue Jan 30 13:24:08 2018 -0600

    Update README.md

commit fe4d4a64f3781304433a4ca2ce3c3037f9eadc6a
Author: Akhil Sangem <akhilsangem@gmail.com>
Date:   Tue Jan 30 13:23:27 2018 -0600

    Create README.md

commit fb2abd8c672d920ab855bc6a93209004024bacb7
Author: Akhil Sangem <akhilsangem@gmail.com>
Date:   Mon Jan 29 23:28:34 2018 -0600

    added urls to notes

commit 5ce8e6042aa69eec278883f4a7380497cccac69a
Author: Akhil Sangem <akhilsangem@gmail.com>
Date:   Mon Jan 29 23:09:13 2018 -0600

    Preparing tcp socket

commit e87823afeeb3ccc672bd6836187984d23180df8a
Author: Akhil Sangem <akhilsangem@gmail.com>
Date:   Mon Jan 29 22:16:51 2018 -0600

    changed usleep

commit f70695e31d8694230868221954a4f27b1cfd6f8e
Author: Akhil Sangem <akhilsangem@gmail.com>
Date:   Mon Jan 29 22:07:43 2018 -0600

    added pthread_join

commit 4408b0afbe68cb3e53791d046e8729174d0c1b1f
Author: Akhil Sangem <akhilsangem@gmail.com>
Date:   Mon Jan 29 22:05:26 2018 -0600

    added pthread.h

commit 576df10b7141cc56e65c40089263e8fa0050b453
Author: Akhil Sangem <akhilsangem@gmail.com>
Date:   Mon Jan 29 22:04:07 2018 -0600

    added unistd.h

commit 5f815ff6f2900f2a5e7a4e078b2fa0a7cedb6063
Author: Akhil Sangem <akhilsangem@gmail.com>
Date:   Mon Jan 29 21:58:14 2018 -0600

    Created agent. Working on getting threads to run

commit 09c8179c130843ca74592f151536e851dc5000f8
Author: Akhil Sangem <akhilsangem@gmail.com>
Date:   Mon Jan 29 19:26:05 2018 -0600

    Added code for tcp server

commit 67f0324e60e01540080223d14282c100655c302e
Author: Akhil Sangem <akhilsangem@gmail.com>
Date:   Mon Jan 29 18:41:18 2018 -0600

    created a small program to listen to tcp and print out what is received

commit 172917f2570c2298aa08fa629193555d5d26a640
Author: Akhil Reddy Sangem - student <arsangem@pyrite-n2.cs.iastate.edu>
Date:   Mon Jan 29 14:58:47 2018 -0600

    edited exampleAgent to run on pyrite

commit 6c076509fdcd5cc778808545aac91b156c7e23e0
Author: Akhil Sangem <akhilsangem@gmail.com>
Date:   Mon Jan 29 14:55:41 2018 -0600

    Added example agent

commit 0bdd51ba94202808a33927d0afe81eae9376140c
Author: Akhil Sangem <akhilsangem@gmail.com>
Date:   Mon Jan 29 14:53:44 2018 -0600

    test

commit 5e386fecf3c76e7a7fc768403e0c760e1ed12275
Author: Akhil Sangem <akhilsangem@gmail.com>
Date:   Mon Jan 29 14:45:03 2018 -0600

    Initial commit


hw1/agent.c

#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <time.h>
#include <netinet/in.h>
#include <unistd.h>
#include <pthread.h>
#include <strings.h>
#include <sys/utsname.h>

void *heartBeat(void *arg){
  typedef struct BEACON
  {
    int ID; // randomly generated during startup
    int CmdPort; // the client listens to this port for manager cmd
    int StartUpTime; // the time when the client starts
    char IP[4]; // the IP address of this client

  } beacon;

  srand((unsigned) time(NULL));
  beacon b;
  b.ID = rand() % 100;
  b.CmdPort = 9002;
  b.StartUpTime = (int)time(NULL);
  b.IP[0] = '0';
  b.IP[1] = '0';
  b.IP[2] = '0';
  b.IP[3] = '0';

  printf("ID is %d\n", b.ID);

  int sock, length, fromlen, n;

  struct sockaddr_in server;
  struct sockaddr_in from;


  int size = sizeof(b);
  char buf[size];

  sock=socket(AF_INET, SOCK_DGRAM, 0);
  if (sock < 0)
  {
    printf("Opening socket\n");
  }

  length = sizeof(server);

  bzero(&server,length);

  server.sin_family=AF_INET;
  server.sin_addr.s_addr=INADDR_ANY;
  server.sin_port=htons(9002);

  fromlen = sizeof(struct sockaddr_in);

  memset(buf, 0x00, size);
  memcpy(buf, &b, size);

  while(1)
  {
    usleep(1);

    //n = sendto(sock,"Got your message\n",17,0, (struct sockaddr *)&from,fromlen);
    n = sendto(sock,buf,sizeof(buf),0, (struct sockaddr *)&server,fromlen);
    while (n < 0)
    {
      printf("sendto failed\n");
      n = sendto(sock,buf,17,0, (struct sockaddr *)&server,fromlen);
    }

  }
  close(sock);
  return NULL;
}

int main(int argc, char const *argv[], char * envp[])
{

  pthread_t beaconHeart;
  pthread_create(&beaconHeart,NULL,heartBeat,NULL);

  struct utsname unameData;
  uname(&unameData);

  char server_message[1];
  char server_response[256] = "Holy Shit you got to the server";

  //create server socket
  int server_socket;
  server_socket = socket(AF_INET, SOCK_STREAM, 0);

  //define server server_address
  struct sockaddr_in server_address;
  server_address.sin_family = AF_INET; //sets type of address
  server_address.sin_port = htons(9004); //sets port
  server_address.sin_addr.s_addr = INADDR_ANY; //connects to localhost

  //bind the socket to our specified ip and sin_port
  bind(server_socket, (struct sockaddr*)  &server_address, sizeof(server_address));

  int client_socket;

  while (1)
  {
    usleep(1);
    printf("Waiting for tcp request\n");
    listen(server_socket, 5);

    client_socket = accept(server_socket, NULL, NULL);
    //bind(client_socket, (struct sockaddr*)  &server_address, sizeof(server_address));
    //bzero(server_message,256);
    recv(client_socket, server_message, sizeof(server_message), 0);

    //printf("Enter 2 to receive OS data\n");
    if(server_message[0] == '2')
    {
      write(client_socket, unameData.sysname, 200);
    }

    write(client_socket, "You Have Connected to the C Server\n", 256);

    printf("%s\n", server_message);

    //close(client_socket);
  }
  //close(server_socket);
  pthread_join(beaconHeart,NULL);
  return 0;
}


cTCPTutorial.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/socket.h>

#include <netinet/in.h>

//simple socket program to print out any data received by tcp

int main(int argc, char const *argv[]) {
  //create socket
  int network_socket;
  //domain, type of socket, specifies protocol
  network_socket = socket(AF_INET, SOCK_STREAM, 0);

  //connect to another socket netinet has a structure for address and remote port

  struct sockaddr_in server_address;
  server_address.sin_family = AF_INET; //sets type of address
  server_address.sin_port = htons(9004); //sets port
  server_address.sin_addr.s_addr = INADDR_ANY; //connects to localhost

  //socket, server address, size of the address, returns a success int
  int connection_status = connect(network_socket, (struct sockaddr *) &server_address, sizeof(server_address));

  //check for error with connection
  if (connection_status == -1) {
    printf("There was an error making a connection to the remote socket \n\n", );
  }

  //call receive function to receive data
  char server_response[256]; //just made a variable to store the data we got back from the receive
  recv(network_socket, &server_response, sizeof(server_response), 0);

  //printout the data that we just got
  printf("The server sent the data: %s\n", server_response);

  //closing socket
  close(network_socket);

//-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/-/
  //now server stuff
  char server_message[256] = "You have reached the server";

  //create server socket
  int server_socket;
  server_socket = socket(AF_INET, SOCK_STREAM, 0);

  //define server server_address
  struct sockaddr_in server_address;
  server_address.sin_family = AF_INET; //sets type of address
  server_address.sin_port = htons(9004); //sets port
  server_address.sin_addr.s_addr = INADDR_ANY; //connects to localhost

  //bind the socket to our specified ip and sin_port
  bind(server_socket, (struct sockaddr*)  &server_address, sizeof(server_address));

  listen(server_socket, 5);

  //holds client socket
  int client_socket;
  client_socket = accept(server_socket, NULL, NULL);

  //send
  send(client_socket, server_message, sizeof(server_message), 0);
  close(server_socket);

  return 0;
}


hw1/hostsender.c
/* Creates a datagram server. The port number is passed as an argument. This server runs forever */
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <time.h>
#include <netinet/in.h>
#include <unistd.h>
#include <strings.h>

void error(char *msg)
{
  perror(msg);
  exit(0);
}

int main(int argc, char *argv[])
{
  int sock, length, fromlen, n;
  struct sockaddr_in server;
  struct sockaddr_in from;
  char buf[1024];

  if (argc < 2)
  {
  fprintf(stderr, "ERROR, no port provided\n");
  exit(0);
  }

  sock=socket(AF_INET, SOCK_DGRAM, 0);
  if (sock < 0)
  {
  error("Opening socket");
  }
  length = sizeof(server);
  bzero(&server,length);
  server.sin_family=AF_INET;
  server.sin_addr.s_addr=inet_addr("127.0.0.1");
  server.sin_port=htons(9002);
  if (bind(sock,(struct sockaddr *)&server,length)<0)
  {
    error("binding");
  }
  fromlen = sizeof(struct sockaddr_in);
  while (1)
  {
    n = recvfrom(sock,buf,1024,0,(struct sockaddr *)&from,(socklen_t *) &fromlen);
    if (n < 0)
    {
    error("recvfrom");
    }
    write(1,"Received a datagram: ",21);
    
    write(1,buf,n);
    //n = sendto(sock,"Got your message\n",17,0, (struct sockaddr *)&from,fromlen);
    n = sendto(sock,"Got your message\n",17,0, (struct sockaddr *)&from,fromlen);
    if (n < 0)
    {
      error("sendto");
    }
  }
}


hw1/exampleAgent.c
#include <stdio.h>
#include <string.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <stdlib.h>

int read_one_byte(int client_socket, char *buffer);
int receiveFully(int client_socket, char *buffer, int length);
void printBinaryArray(char *buffer, int length);
int toInteger32(char *bytes);
void convertUpperCase(char *buffer, int length);

int main()
{
    // get a socket descriptor
   	int server_socket = socket(PF_INET, SOCK_DGRAM, IPPROTO_UDP);
   	printf("server_socket = %d\n", server_socket);

    // bind to a port
 	struct sockaddr_in sin;
    memset(&sin, 0, sizeof(sin));
    //sin.sin_len = sizeof(sin);  // comment this line out if running on pyrite (linux)
    sin.sin_family = AF_INET; // or AF_INET6 (address family)
    sin.sin_port = htons(9002);
    sin.sin_addr.s_addr= INADDR_ANY;

    if (bind(server_socket, (struct sockaddr *)&sin, sizeof(sin)) < 0)
    {
        // Handle the error.
		printf("bind error\n");
    }

    listen(server_socket, 5); /* maximum 5 connections will be queued */
	int counter = 0;
	while (1)
	{
		struct sockaddr client_addr;
		unsigned int client_len;

		printf("accepting ....\n");
  		int client_socket = accept(server_socket, &client_addr, &client_len);
  		printf("request %d comes ...\n", counter++);
		/* launch a new thread to take care of this client connection */
		/* client_addr contains the address of the connecting client */
		/* client_len is the buffer length that is valid in client_addr */
		/* both client_addr and client_len are optional */

		// processing this request

		// get the command length
		char packet_length_bytes[4];
		//receiveFully(client_socket, packet_length_bytes, 4);

		printBinaryArray(packet_length_bytes, 4);

		int packet_length = toInteger32(packet_length_bytes);
		printf("packet_length_bytes = %d", packet_length);

		// allocate buffer to receive the data
		char *buffer = (char *)malloc(packet_length);
		//receiveFully(client_socket, buffer, packet_length);

		// convert to upper case
	    convertUpperCase(buffer, packet_length);

	    // send back
	    send(client_socket, packet_length_bytes, 4, 0); // 4 bytes first
	    send(client_socket, buffer, packet_length, 0);

	    // release buffer
	    free(buffer);
	}
}

void convertUpperCase(char *buffer, int length)
{
	int i = 0;
	while (i < length)
	{
		if (buffer[i] >= 'a' && buffer[i] <= 'z')
		{
			buffer[i] = buffer[i] - 'a' + 'A';
		}
		i++;
	}
}

int receive_one_byte(int client_socket, char *cur_char)
{
    ssize_t bytes_received = 0;
	while (bytes_received != 1)
	{
		bytes_received = recv(client_socket, cur_char, 1, 0);
	}

	return 1;
}

int receiveFully(int client_socket, char *buffer, int length)
{
	char *cur_char = buffer;
	ssize_t bytes_received = 0;
	while (bytes_received != length)
	{
	    receive_one_byte(client_socket, cur_char);
	    cur_char++;
	    bytes_received++;
	}

	return 1;
}

void printBinaryArray(char *buffer, int length)
{
    int i=0;
    while (i<length)
    {
		printf("%d ", buffer[i]);
		i++;
	}

	printf("\n");
}

int toInteger32(char *bytes)
{
	int tmp = (bytes[0] << 24) +
	          (bytes[1] << 16) +
	          (bytes[2] << 8) +
	          bytes[3];

	return tmp;
}

/* On an OS with a different endian from java, use this one
int toInteger32(char *bytes)
{
	int tmp = bytes[0] +
	          (bytes[1] << 8) +
	          (bytes[2] << 16) +
	          (bytes[3] << 24);

	return tmp;
}
*/
